Documentation update: Forge2D (#733)
flame-engine/flame: doc/forge2d.md

# Forge2D

> We (the Flame organization) maintains a ported version of the Box2D physics engine and our version

Reviewer: ```suggestion
We (the Flame organization) maintain a ported version of the Box2D physics engine and our version
```

is called Forge2D.

If you want to use Forge2D specifically for Flame you should use our bridge library
[flame_forge2d](https://github.com/flame-engine/flame_forge2d) and if you just want to use it
in a Dart project you can use the [forge2d](https://github.com/flame-engine/forge2d) library
directly.

> To use it in your game you just need to add flame_forge2d to your pubspec.yaml, as can be seen

Reviewer: ```suggestion
To use it in your game you just need to add `flame_forge2d` to your pubspec.yaml, as can be seen
```

in the [Forge2D example](https://github.com/flame-engine/flame_forge2d/tree/main/example) and in the
pub.dev [installation instructions](https://pub.dev/packages/flame_forge2d).

## Forge2DGame (BaseGame extension)

If you are going to use Forge2D in your project it can be a good idea to use the Forge2D specific
extension of the `BaseGame` class.

> It is called `Forge2DGame` and it will control the adding and removal of Forge2D's BodyComponents

Reviewer: ```suggestion
It is called `Forge2DGame` and it will control the adding and removal of Forge2D's `BodyComponents`
```

as well as your normal components.

If you are previously familiar with Box2D it can be good to know that the whole concept of the

> box2d world is mapped to `world` in the `Forge2DGame` component and every Body should be a

Reviewer: ```suggestion
Box2d world is mapped to `world` in the `Forge2DGame` component and every `Body` should be a
```

`BodyComponent`, and added to your `Forge2DGame`.

> You can have for example a HUD and other non-physics-related components in your `Forge2DGame's

Reviewer: ```suggestion
You can have for example a HUD and other non-physics-related components in your `Forge2DGame`'s
```

component list along with your physical entities. When the update is called, it will use the Forge2D
physics engine to properly update every child.

A simple `Forge2DGame` implementation examples can be seen in the
[examples folder](https://github.com/flame-engine/flame_box2d/blob/main/examples/).

## BodyComponent

> If you don't need to have a sprite on top of your body you should use the plain BodyComponent, for

Reviewer: ```suggestion
If you don't need to have a sprite on top of your body you should use the plain `BodyComponent`, for
```

example if you want a circle, rectangle or polygon but only painted with a Flutter `Paint`.

## SpriteBodyComponent

> Often you want to render a sprite on top of the BodyComponent that you are going to use in your

Reviewer: ```suggestion
Often you want to render a sprite on top of the `BodyComponent` that you are going to use in your
```

> Forge2DGame. This component will handle the scaling and positioning of your sprite on top of the

Reviewer: ```suggestion
`Forge2DGame`. This component will handle the scaling and positioning of your sprite on top of the
```

body.

## PositionBodyComponent

One of the most commonly used classes in Flame is the `PositionComponent`, many of the commonly used
components in Flame are subclasses of `PositionComponent`. If you want to put a `PositionComponent`
or any of its subclasses on top of a Forge2D body you can use the `PositionBodyComponent` and it
will, just like with the `SpriteBodyComponent`, handle the rotation, positioning and scaling of that
component so that it follows the underlying `Body`.

## Contact callbacks

If you are using `Forge2DGame` you can take advantage of its way of handling contacts between two
`BodyComponent`s.

When creating the body definition for your `BodyComponent` make sure that you set the userdata to
the current object, otherwise it will not be possible to detect collisions.
Like this:
```dart
final bodyDef = BodyDef()
  // To be able to know which component that is involved in a collision
  ..setUserData(this);
```

Now you have to make an implementation of `ContactCallback` where you set which two types that it
should react when they come in contact.

> If you have two `BodyComponent`s named Ball and Wall and you want to do something when they come in

Reviewer: ```suggestion
If you have two `BodyComponent`s named `Ball` and `Wall` and you want to do something when they come in
```

> contact you would do like this:

Reviewer: ```suggestion
contact, you could do something like this:
```
